src/timeseriesUpdate.ts

```diff
@@ -0,0 +1,60 @@
+import {Address, BigInt, log} from '@graphprotocol/graph-ts';
+import { Notional, UpdateAssetRate, UpdateETHRate } from '../generated/Notional/Notional';
+
+import { 
+    AssetExchangeRate,
+    EthExchangeRate
+} from '../generated/schema';
+
+import { 
+    getAssetExchangeRateHistoricalData,
+    getEthExchangeRateHistoricalData
+} from './exchange_rates/utils'
+
+/* 
+    Historical data is stored on an hourly basis to keep it lean, this could be changed in the future
+    or be kept as multiple time periods. New values during the timespan updates the hourly value
+*/
+
+function createHourlyId(currencyId: number, timestamp: i32): string {
+    let uniqueHourIndex = timestamp / 3600; // Integer division will always floor result
+
+    return currencyId
+        .toString()
+        .concat(':')
+        .concat(uniqueHourIndex.toString());
+}
+
+export function updateEthExchangeRateHistoricalData(ethExchangeRate: EthExchangeRate, event: UpdateETHRate): void {
+    let historicalId = createHourlyId(event.params.currencyId, event.block.timestamp.toI32());
+    let ethExchangeRateHistoricalData = getEthExchangeRateHistoricalData(historicalId);
+    let roundedTimestamp = (event.block.timestamp.toI32() / 3600) * 3600;
+
+    let notional = Notional.bind(event.address);
+    let rateResult = notional.getCurrencyAndRates(event.params.currencyId);
+    let ethRate = rateResult.value2;
+
+    ethExchangeRateHistoricalData.timestamp = roundedTimestamp;
+    ethExchangeRateHistoricalData.value = ethRate.rate.div(BigInt.fromI32(10).pow(10));
+    ethExchangeRateHistoricalData.ethExchangeRate = ethExchangeRate.id;
+
+    log.debug('Updated ethExchangeRateHistoricalData variables for entity {}', [ethExchangeRateHistoricalData.id]);
+    ethExchangeRateHistoricalData.save();
+}
+
+export function updateAssetExchangeRateHistoricalData(assetExchangeRate: AssetExchangeRate, event: UpdateAssetRate): void {
+    let historicalId = createHourlyId(event.params.currencyId, event.block.timestamp.toI32());
+    let assetExchangeRateHistoricalData = getAssetExchangeRateHistoricalData(historicalId);
+    let roundedTimestamp = (event.block.timestamp.toI32() / 3600) * 3600;
+
+    let notional = Notional.bind(event.address);
+    let rateResult = notional.getCurrencyAndRates(event.params.currencyId);
```

Comment:
As a consequence of calling this method via a blockHandler, you won't have access to the `event` object anymore. What I would recommend for getting the Notional address and the currencyId is this:

1. for the Notional address you can get it via `dataSource.address()` https://thegraph.com/docs/en/developer/assemblyscript-api/#data-source-metadata
2. for the currency id, you can call Notional.getMaxCurrencyId() and loop through currency id values where `1 <= currencyId <= maxCurrencyId` and get the data for each currencyId.

We should also store these two values:
notional.nTokenPresentValueAssetDenominated(currencyId)
notional.nTokenPresentValueUnderlyingDenominated(currencyId)

These methods may revert if there is no nToken present. In this case you do not need to store any value. You can wrap these calls in a try catch like this:
```
let pvAsset = notional.try_nTokenPresentValueAssetDenominated(currencyId)
let pvUnderlying = notional.try_nTokenPresentValueUnderlyingDenominated(currencyId)
if (pvAsset.reverted || pvUnderlying.reverted) return;
```

https://github.com/notional-finance/contracts-v2/blob/master/contracts/external/actions/nTokenAction.sol#L217-L233